src/viser/_gui_api.py

```diff
@@ -498,6 +505,81 @@ def add_gui_markdown(
         handle.content = content
         return handle
 
+    def add_gui_plotly(
+        self,
+        figure: go.Figure,
+        aspect_ratio: float = 1.0,
+        order: Optional[float] = None,
+        visible: bool = True,
+        font: Optional[str] = None,
+    ) -> GuiPlotlyHandle:
+        """Add a Plotly Plot to the GUI.
+
+        Args:
+            figure: Plotly figure to display.
+            aspect_ratio: Aspect ratio of the plot in the control panel (width / height).
+            order: Optional ordering, smallest values will be displayed first.
+            visible: Whether the component is visible.
+            font: Optional font to use for the plot.
+
+        Returns:
+            A handle that can be used to interact with the GUI element.
+        """
+        handle = GuiPlotlyHandle(
+            _gui_api=self,
+            _id=_make_unique_id(),
+            _visible=visible,
+            _parent_container_id=self._get_container_id(),
+            _order=_apply_default_order(order),
+            _figure=figure,
+            _aspect_ratio=aspect_ratio,
+            _font=font,
+        )
+
+        # If plotly.min.js hasn't been sent to the client yet, the client won't be able
+        # to render the plot. Send this large file now! (~3MB)
+        if not self._setup_plotly_js:
+            # Check if plotly is installed.
+            try:
+                import plotly
+            except ImportError:
+                raise ImportError(
+                    "You must have the `plotly` package installed to use the Plotly GUI element."
+                )
+
+            # Check that plotly.min.js exists.
+            plotly_path = (
+                Path(plotly.__file__).parent / "package_data" / "plotly.min.js"
+            )
+            assert (
+                plotly_path.exists()
+            ), f"Could not find plotly.min.js at {plotly_path}."
+
+            # Send it over!
+            with open(plotly_path, "r") as f:
+                plotly_js = f.read()
+            self._get_api()._queue(
+                _messages.RunJavascriptMessage(
+                    source=plotly_js,
+                )
+            )
+
+            # Update the flag so we don't send it again.
+            self._setup_plotly_js = True
+
+        # After plotly.min.js has been sent, we can send the plotly figure.
+        self._get_api()._queue(
+            _messages.GuiAddPlotlyMessage(
+                order=handle._order,
+                id=handle._id,
+                plotly_json_str=handle.plot_to_json(),
```

Comment:
what do you think about following the same pattern as `add_gui_markdown()`: https://github.com/nerfstudio-project/viser/blob/d11f9b9186d64171c262c2509596d461b6c060e2/src/viser/_gui_api.py#L493-L505

where attributes are set to their desired values only after the initial `GuiAddXXXMessage` is sent, using the setters exposed in the `handle` object

this is mostly a robustness nit: since we don't have unit tests, it seems nice to use internally what we expose externally, reduce redundant ways of doing the same thing, etc (even if one is less efficient)

Reply:
Ok, I tried to update the code to rely on the `GuiAddXXXMessage`s, but not sure if this implementation was what you intended. 

https://github.com/nerfstudio-project/viser/blob/3af36b1bb606bbb26bd59737c4d44eb6333cecc7/src/viser/_gui_api.py#L571-L587